_posts/2018-11-01-http-part6.markdown

```diff
@@ -0,0 +1,514 @@
+     * it is assumed that connection is no longer active. It is optimization so we don't have to call time()
+     * for every byte on socket.
+     */
+    // (3)
+    private val inactiveChecksThreshold = (keepAliveTimeout.toMillis() / checkKeepAliveInterval.toMillis()).toInt()
+
+    internal fun isInactive(): Boolean = checks++ > inactiveChecksThreshold // (2)
+
+    fun close() {
+        try {
+            socket.close()
+        } catch (e: Exception) {
+            logger.warn("Could not close the connection", e)
+        }
+    }
+
+    fun isClosed(): Boolean = socket.isClosed
+
+    fun getInputStream(): InputStream = socket.getInputStream()
+            ?.let(::ActivityTrackingInputStream)
+            ?: throw ConnectionException("Could not obtain stream. Is socket closed?")
+
+    fun getOutputStream(): OutputStream = socket.getOutputStream()
+            ?.let(::ActivityTrackingOutputStream)
+            ?: throw ConnectionException("Could not obtain stream. Is socket closed?")
+
+    private inner class ActivityTrackingInputStream(val inputStream: InputStream): InputStream() {
+
+        override fun read(): Int = inputStream.read().also { checks = 0 }
+        override fun read(b: ByteArray?): Int = inputStream.read(b).also { checks = 0 }
+        override fun read(b: ByteArray?, off: Int, len: Int): Int = inputStream.read(b, off, len).also { checks = 0 }
+        override fun skip(n: Long): Long = inputStream.skip(n)
+        override fun available(): Int = inputStream.available()
+        override fun reset() = inputStream.reset()
+        override fun close() = inputStream.close()
+        override fun mark(readlimit: Int) = inputStream.mark(readlimit)
+        override fun markSupported(): Boolean = inputStream.markSupported()
+    }
+
+    private inner class ActivityTrackingOutputStream(val outputStream: OutputStream): OutputStream() {
+
+        override fun write(b: Int) = outputStream.write(b).also { checks = 0 }
+        override fun write(b: ByteArray?) = outputStream.write(b).also { checks = 0 }
+        override fun write(b: ByteArray?, off: Int, len: Int) = outputStream.write(b, off, len).also { checks = 0 }
+        override fun flush() = outputStream.flush()
+        override fun close() = outputStream.close()
+    }
+}
+
+open class ConnectionException(msg: String, throwable: Throwable? = null): RuntimeException(msg, throwable)
+```
+How should we tell whether the connection is no longer active? It would be so easy to save the timestamp of the last activity and compare it to `keepAliveTimeout`, but calling `System.nanoTime()` on every byte send/received would create an overhead. Instead, we create a counter (1) that is incremented on every inactivity check (2).  We know how often this method will be called because of the `checkKeepAliveInternal` argument, so we can compute (3) after how many checks the connection is no longer active.
```

Comment:
whether the connection is active or not

Reply:
How *can* we tell whether the connection is no longer active?

How can we tell - we don't know how to do something, and we're asking someone or just posing a rhetorical question.
How should we tell - we know something and we're thinking about the way we should phrase it.